primitiv/c/graph.cc

```diff
@@ -7,224 +7,193 @@
-  SAFE_RETURN(primitiv_Node_operator_id(node), status, 0);
+primitiv_Status primitiv_Node_operator_id(const primitiv_Node *node,
+                                          uint32_t *id) {
+  try {
+    *id = to_cpp_ptr(node)->operator_id();
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
 }
 
-uint32_t primitiv_Node_value_id(const primitiv_Node *node) {
-  return to_cc(node)->value_id();
-}
-uint32_t safe_primitiv_Node_value_id(const primitiv_Node *node,
-                                     primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Node_value_id(node), status, 0);
+primitiv_Status primitiv_Node_value_id(const primitiv_Node *node,
+                                       uint32_t *id) {
+  try {
+    *id = to_cpp_ptr(node)->value_id();
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
 }
 
-primitiv_Shape *primitiv_Node_shape(const primitiv_Node *node) {
-  return to_c_from_value(to_cc(node)->shape());
-}
-primitiv_Shape *safe_primitiv_Node_shape(const primitiv_Node *node,
-                                         primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Node_shape(node), status, nullptr);
+primitiv_Status primitiv_Node_shape(const primitiv_Node *node,
+                                    primitiv_Shape **shape) {
+  try {
+    *shape = to_c_ptr_from_value(to_cpp_ptr(node)->shape());
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
 }
 
-primitiv_Device *primitiv_Node_device(const primitiv_Node *node) {
-  return to_c(&to_cc(node)->device());
-}
-primitiv_Device *safe_primitiv_Node_device(const primitiv_Node *node,
-                                           primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Node_device(node), status, nullptr);
+primitiv_Status primitiv_Node_device(const primitiv_Node *node,
+                                     primitiv_Device **device) {
+  try {
+    *device = to_c_ptr(&to_cpp_ptr(node)->device());
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
 }
 
-float primitiv_Node_to_float(const primitiv_Node *node) {
-  return to_cc(node)->to_float();
-}
-float safe_primitiv_Node_to_float(const primitiv_Node *node,
-                                  primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Node_to_float(node), status, 0.0);
+primitiv_Status primitiv_Node_to_float(const primitiv_Node *node,
+                                       float *value) {
+  try {
+    *value = to_cpp_ptr(node)->to_float();
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
 }
 
-void primitiv_Node_to_array(const primitiv_Node *node, float *array) {
-  std::vector<float> v = to_cc(node)->to_vector();
-  std::copy(v.begin(), v.end(), array);
-}
-void safe_primitiv_Node_to_array(const primitiv_Node *node,
-                                 float *array,
-                                 primitiv_Status *status) {
-  SAFE_EXPR(primitiv_Node_to_array(node, array), status);
+primitiv_Status primitiv_Node_to_array(const primitiv_Node *node,
```

Comment:
This function should take a size of the array and check the boundary, similar to functions that return strings.